src/SignalR/clients/java/signalr/core/src/main/java/com/microsoft/signalr/HubConnection.java

```diff
@@ -517,33 +517,44 @@ private void stopConnection(String errorMessage) {
         RuntimeException exception = null;
         this.state.lock();
         try {
+            ConnectionState connectionState = this.state.getConnectionStateUnsynchronized(true);
+
+            if (connectionState == null)
+            {
+                this.logger.error("'stopConnection' called with a null ConnectionState. This is not expected, please file a bug.");
+                return;
+            }
+
             // errorMessage gets passed in from the transport. An already existing stopError value
             // should take precedence.
-            if (this.state.getConnectionStateUnsynchronized(false).stopError != null) {
-                errorMessage = this.state.getConnectionStateUnsynchronized(false).stopError;
+            if (connectionState.stopError != null) {
+                errorMessage = connectionState.stopError;
             }
             if (errorMessage != null) {
                 exception = new RuntimeException(errorMessage);
                 logger.error("HubConnection disconnected with an error {}.", errorMessage);
             }
 
-            ConnectionState connectionState = this.state.getConnectionStateUnsynchronized(true);
-            if (connectionState != null) {
-                connectionState.cancelOutstandingInvocations(exception);
-                connectionState.close();
-                this.state.setConnectionState(null);
-            }
+            this.state.setConnectionState(null);
+            connectionState.cancelOutstandingInvocations(exception);
+            connectionState.close();
 
             logger.info("HubConnection stopped.");
-            this.state.changeState(HubConnectionState.CONNECTED, HubConnectionState.DISCONNECTED);
+            // We can be in the CONNECTING or CONNECTED state here, depending on if the handshake response was received or not.
+            // connectionState.close() above will exit the Start call with an error if it's still running
+            this.state.changeState(HubConnectionState.DISCONNECTED);
```

Comment:
Given the comment, shouldn't we be changing the state before calling connectionState.close() so that if someone has a retry loop on HubConnection.start() without a delay, it doesn't fail due still being in the CONNECTING state?

Unless the following is want prevents the retry loop from failing:

https://github.com/dotnet/aspnetcore/blob/9b6103989c0a4a2b0ef0a9f430525bfcd46c6be7/src/SignalR/clients/java/signalr/core/src/main/java/com/microsoft/signalr/HubConnection.java#L330-L336

In that case, the problem is even worse. Instead of the retry loop spinning trying to start the connection while its still in the CONNECTING state, it would start connecting again only for this line to incorrectly put the new connection in the DISCONNECTED state because of the old connection closed.

Reply:
> if someone has a retry loop on HubConnection.start() without a delay, it doesn't fail due still being in the CONNECTING state?

There is a lock, so they wont get anywhere.

Reply:
And it's not like a TCS where the continuation and run inline?

Reply:
Which continuation are we talking about? They can run inline.